### PR TITLE
Fix expecttest accept on Python 3.8 and later

### LINKED ISSUE #59705: Failed to substitute string at {fn}:{lineno}; did you use triple quotes? on Python 3.8 and later
Steps to reproduce:
1. Build PyTorch on Python 3.8
2. Update a multiline expect test and try to use EXPECTTEST_ACCEPT=1

It doesn't work

```
======================================================================
FAIL: test_basic (__main__.TestPythonDispatch)                                                               
----------------------------------------------------------------------              
Traceback (most recent call last):
  File "test/test_python_dispatch.py", line 107, in test_basic
    self.assertExpectedInline('\n'.join(logs), '''\                                                          
  File "/data/users/ezyang/pytorch-tmp2/torch/testing/_internal/expecttest.py", line 233, in assertExpectedIn
line
    assert old != new, f"Failed to substitute string at {fn}:{lineno}; did you use triple quotes?"
AssertionError: Failed to substitute string at test/test_python_dispatch.py:107; did you use triple quotes?

----------------------------------------------------------------------
```

It looks like the reason is that Python is now reporting the error to be on the BEGINNING of the line rather than the end of the line, oopsies.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @Varal7

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59760 Dispatch to Python via __torch_dispatch__
* #59758 Make PyObject_FastGetAttrString accept const char*
* **#59709 Fix expecttest accept on Python 3.8 and later**

Fixes #59705.

Python 3.8 fixed tracebacks to report the beginning of the line
that raised an error, rather than the end.  This makes for a simpler
implementation (no more string reversing) but need to actually
implement.  This wasn't caught by tests because we hard coded line
numbers to do substitutions, so I also added a little smoketest to
detect future changes to traceback line number behavior.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28994919](https://our.internmc.facebook.com/intern/diff/D28994919)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28994919).

**Comment by ezyang** [comment]
Note to reviewers, I don't think either of you are experts in this code, I hope you can just review it on its merits.

File: torch/testing/_internal/expecttest.py [hunk 1]
```diff
@@ -152,54 +168,85 @@ def replace_string_literal(src : str, lineno : int,
...
+
+            return ''.join([m.group('prefix'),
+                            'r' if raw else '',
+                            m.group('quote'),
+                            new_body,
+                            m.group('quote'),
+                            ])
+
+        return (src[:i] + RE_EXPECT.sub(replace, src[i:], count=1), delta[0])
+    else:
```
Comment (bdhirsh) [inline torch/testing/_internal/expecttest.py:221]: more of these two conditions could probably be de-duped? Not a big deal though

**Review by bdhirsh — APPROVED** [review]

File: test/test_expecttest.py [hunk 2]
```diff
@@ -96,6 +99,72 @@ def test_sample(self):
 ''')
 """)
 
+    def test_sample_lineno_at_start(self):
```
Comment (walterddr) [inline test/test_expecttest.py:102]: nit: 
maybe specifically named previous test in line:49 as 
`test_sample_lineno_at_end(self)`

**Review by walterddr — APPROVED** [review]

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28994919).

**Pull request closed without merging**